File: 滑动窗口/239.滑动窗口的最大值.py
from collections import deque
from typing import List

# k > 0  nums.length > 0

class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        res = []
        deq = deque()
        for l in range(len(nums)):
            while deq and deq[0] < l - k + 1:  # 移除窗口外的数的索引
                deq.popleft()
            # 实现单调队列, 维护左边的数字比右边大
            while deq and nums[deq[-1]] < nums[l]:
                deq.pop()
            deq.append(l)
            # 窗口形成后, 每次都取最大值
            if l + 1 >= k:
                res.append(nums[deq[0]])
        return res

File: 滑动窗口/239.滑动窗口的最大值.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
var maxSlidingWindow = function (nums, k) {
  //! q存放的是索引
  const q = [], res = [];
  for (let i = 0; i < nums.length; i++) {
    //* 维护窗口大小限制
    while (q.length && q[0] < i - k + 1) q.shift();
    //* 维护单调性
    while (q.length && nums[i] > nums[q[q.length - 1]]) q.pop();
    q.push(i);
    //* 将结果加入数组
    if (i + 1 >= k) res.push(nums[q[0]]);
  }
  return res;
};
